### DISCUSSION TITLE
How to procedurally generate pre-built rooms without hallways? And how to save pre-built rooms and use them?

### DISCUSSION THREAD

**Discussion by ghost** [Minestom/Minestom#1797] [Help]
The title says it all.

**Comment by reeve567** [comment]
This is a big question, so I'll only be able to really give you tips here.
Assuming you want to do some kind of dungeon style system, I would look into different kinds of maze algorithms, such as recursive backtracking. From here you'd have yourself some kind of maze data structure. I have a Kotlin library that generates mazes with these algorithms, but [this](https://www.amazon.com/Mazes-Programmers-Twisty-Little-Passages/dp/1680500554) book covers it really well.
Assuming you want this to be in a grid shape style, you'll have a shape of cells, where each cell will have north, west, south, east.

Once you've got that down, decide how you want to store these rooms, whether that be the typical schema format in files or in a database of some kind. I did something like this before, (albeit before I knew about the maze algorithms) except instead of using schemas I just built the rooms in some random chunks in another world and copied the chunks over.

The main kind of rooms you'll need:
- `Deadend` room
- `Hallway` room - A room where there's an exit on opposite sides
- `Corner` room - As you can probably guess, a room with 2 exits next to each other
- `T` room - 3 exits
- `Intersection` - 4 exits

Depending on how you set this up, you might need to create your rotations of these rooms ahead of time, or rotate them before they're placed, through code.
As far as placing goes, that comes down to your format and the solution you come up with, afaik there are quite a few different people that have implemented schemas in Minestom.
